File: frontend-web/src/app/core/services/pagos.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { environment } from '../../../environments/environment';
import { Pago, PaymentIntentCreate, PaymentIntentResponse } from '../models/pago.model';

@Injectable({ providedIn: 'root' })
export class PagosService {
  private http = inject(HttpClient);
  private readonly base = `${environment.apiUrl}/pagos`;

  crearIntent(data: PaymentIntentCreate) {
    return this.http.post<PaymentIntentResponse>(`${this.base}/crear-intent`, data);
  }

  confirmar(paymentIntentId: string) {
    return this.http.post<Pago>(`${this.base}/confirmar`, { payment_intent_id: paymentIntentId });
  }

  listarPorVenta(ventaId: number) {
    return this.http.get<Pago[]>(`${this.base}/venta/${ventaId}`);
  }

  crearPagoQR(ventaId: number) {
    return this.http.post<Pago>(`${this.base}/qr`, { venta_id: ventaId });
  }

  obtenerPago(pagoId: number) {
    return this.http.get<Pago>(`${this.base}/${pagoId}`);
  }

  confirmarPagoQR(pagoId: number, aprobar: boolean, motivoRechazo?: string) {
    return this.http.patch<Pago>(`${this.base}/${pagoId}/confirmar-qr`, {
      aprobar,
      motivo_rechazo: motivoRechazo,
    });
  }

  listarPendientesVerificacion(sucursalId?: number) {
    let params = new HttpParams();
    if (sucursalId) params = params.set('sucursal_id', sucursalId);
    return this.http.get<Pago[]>(`${this.base}/pendientes-verificacion`, { params });
  }
}

File: mobile-app/lib/features/checkout/services/pagos_service.dart
import 'dart:typed_data';
import 'package:dio/dio.dart';
import 'package:image_picker/image_picker.dart';
import '../../../core/network/dio_client.dart';
import '../models/pago.dart';

class PagosService {
  final DioClient _client = DioClient();

  Future<PaymentIntentResponse> crearPaymentIntent({
    required int ventaId,
    String metodo = 'tarjeta',
  }) async {
    final response = await _client.post<Map<String, dynamic>>(
      '/pagos/crear-intent',
      data: PaymentIntentCreate(ventaId: ventaId, metodo: metodo).toJson(),
    );

    if (response.data == null) {
      throw Exception('Error al obtener respuesta de payment intent');
    }

    return PaymentIntentResponse.fromJson(response.data!);
  }

  Future<Pago> confirmarPago({required String paymentIntentId}) async {
    final response = await _client.post<Map<String, dynamic>>(
      '/pagos/confirmar',
      data: {'payment_intent_id': paymentIntentId},
    );

    if (response.data == null) {
      throw Exception('Error al confirmar pago');
    }

    return Pago.fromJson(response.data!);
  }

  Future<Pago> crearPagoQR({required int ventaId}) async {
    final response = await _client.post<Map<String, dynamic>>(
      '/pagos/qr',
      data: {'venta_id': ventaId},
    );

    if (response.data == null) {
      throw Exception('Error al generar pago QR');
    }

    return Pago.fromJson(response.data!);
  }

  Future<Pago> subirComprobante({
    required int pagoId,
    required XFile archivo,
  }) async {
    final bytes = await archivo.readAsBytes();
    final formData = FormData.fromMap({
      'archivo': MultipartFile.fromBytes(
        bytes,
        filename: archivo.name.isNotEmpty ? archivo.name : 'comprobante.jpg',
      ),
    });

    final response = await _client.post<Map<String, dynamic>>(
      '/pagos/$pagoId/comprobante',
      data: formData,
    );

    if (response.data == null) {
      throw Exception('Error al subir comprobante');
    }

    return Pago.fromJson(response.data!);
  }

  Future<Pago> subirComprobanteBytes({
    required int pagoId,
    required Uint8List bytes,
    String nombreArchivo = 'comprobante.jpg',
  }) async {
    final formData = FormData.fromMap({
      'archivo': MultipartFile.fromBytes(
        bytes,
        filename: nombreArchivo,
      ),
    });

    final response = await _client.post<Map<String, dynamic>>(
      '/pagos/$pagoId/comprobante',
      data: formData,
    );

    if (response.data == null) {
      throw Exception('Error al subir comprobante');
    }

    return Pago.fromJson(response.data!);
  }

  Future<List<Pago>> obtenerPagoDeVenta(int ventaId) async {
    final response = await _client.get<List<dynamic>>('/pagos/venta/$ventaId');
    final list = response.data ?? [];
    return list.map((item) => Pago.fromJson(item as Map<String, dynamic>)).toList();
  }

  Future<Pago> obtenerPago(int pagoId) async {
    final response = await _client.get<Map<String, dynamic>>('/pagos/$pagoId');
    if (response.data == null) {
      throw Exception('Pago no encontrado');
    }
    return Pago.fromJson(response.data!);
  }
}
